### ISSUE TITLE
MEMPTR not marked as volatile

### ISSUE THREAD

**Issue by asiekierka** [iss000/oricCompilerBenchmark#6]
In `libmos6502vm/types.h`:

```c
#ifndef __HOST_C__
#define MEMPTR(address)       ((uint8_t*)(address))
```

Because `address` is not marked as volatile, the compiler is allowed by the C standard to elide (remove) multiple writes to it. This causes the `0xcafe` benchmark to not pass on LLVM-MOS, because it gets converted from C to the following LLVM IR:

```
define dso_local i16 @main() local_unnamed_addr #0 {
entry:
  tail call void @_puts(ptr noundef nonnull @.str) #2
  tail call void @_putc(i8 noundef zeroext 67) #2
  tail call void @_putc(i8 noundef zeroext 65) #2
  tail call void @_putc(i8 noundef zeroext 70) #2
  tail call void @_putc(i8 noundef zeroext 69) #2
  tail call void @_puts(ptr noundef nonnull @.str.1) #2
  tail call void @_putc(i8 noundef zeroext 10) #2
  ret i16 0
}
```

These serial calls to `_putc` are then inlined further into writes to memory location `CPORT`, but as they are not *volatile* writes, the C compiler may assume it is legal to remove duplicate writes into the same memory address if there are no reads happening from that address. This is why the code then returns `0XCAFE == 0XE :)`.

The proposed fix is to adjust this to:

```c
#ifndef __HOST_C__
#define MEMPTR(address)       ((volatile uint8_t*)(address))
```

**Comment by asiekierka** [comment]
To clarify, this fixes the `0xcafe` and `aes256` benchmarks for `llvm-mos`; I have not tested other compilers, as I don't have all of them set up.

**Comment by iss000** [comment]
Thanks, @asiekierka! Fixed.

**Comment by asiekierka** [comment]
Note that SDCC also uses `volatile` (at least I've seen it being used in the GBDK toolchain). It's part of the C standard, after all.

**Comment by asiekierka** [comment]
(Also, just out of curiosity, what's with all of the custom LDCOPTS like `-mllvm -force-precise-rotation-cost`? We don't use them in mainline llvm-mos-sdk; if they actually provably help, we'd like to know about this, and if they don't, it might be best to keep as close to llvm-mos-sdk's official setup as possible, to have a more representative result for end users.

If you want to use our official compiler runtime and such, you might be better off using `mos-common-clang` from `llvm-mos-sdk` and adding your own link script on top of it. That would fix #7 too.)
